service/frontend/nexus_incoming_service_client.go

```diff
@@ -0,0 +1,373 @@
+) (*operatorservice.GetNexusIncomingServiceResponse, error) {
+	if err := validateGetRequest(request); err != nil {
+		return nil, err
+	}
+
+	entry, err := c.persistence.GetNexusIncomingService(ctx, &p.GetNexusIncomingServiceRequest{
+		ServiceID: request.Id,
+	})
+	if err != nil {
+		var notFound *serviceerror.NotFound
+		if errors.As(err, &notFound) {
+			return nil, err
+		}
+		c.logger.Error(fmt.Sprintf("error looking up Nexus incoming service with ID `%v` from persistence", request.Id), tag.Error(err))
+		return nil, serviceerror.NewInternal(fmt.Sprintf("error looking up Nexus incoming service with ID `%v`", request.Id))
+	}
+
+	return &operatorservice.GetNexusIncomingServiceResponse{
+		Service: persistedEntryToExternalAPI(entry),
+	}, nil
+}
+
+func (c *NexusIncomingServiceClient) List(
+	ctx context.Context,
+	request *operatorservice.ListNexusIncomingServicesRequest,
+) (*operatorservice.ListNexusIncomingServicesResponse, error) {
+	if request.GetName() != "" {
+		return c.listAndFilterByName(ctx, request)
+	}
+
+	pageSize := request.GetPageSize()
+	if pageSize == 0 {
+		pageSize = int32(c.config.listDefaultPageSize())
+	} else if err := c.validatePageSize(pageSize); err != nil {
+		return nil, err
+	}
+
+	resp, err := c.persistence.ListNexusIncomingServices(ctx, &p.ListNexusIncomingServicesRequest{
+		LastKnownTableVersion: 0,
+		NextPageToken:         request.NextPageToken,
+		PageSize:              int(pageSize),
+	})
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("error listing Nexus incoming services from persistence. NextPageToken: %v PageSize: %d", request.NextPageToken, pageSize), tag.Error(err))
+		return nil, serviceerror.NewInternal("error listing Nexus incoming services")
+	}
+
+	services := make([]*nexus.IncomingService, len(resp.Entries))
+	for i, entry := range resp.Entries {
+		services[i] = persistedEntryToExternalAPI(entry)
+	}
+
+	return &operatorservice.ListNexusIncomingServicesResponse{
+		NextPageToken: resp.NextPageToken,
+		Services:      services,
+	}, nil
+}
+
+// listAndFilterByName paginates over all services returned by persistence layer to find the service name
+// indicated in the request. Returns that service if found or an empty response if not.
+// PageSize and NextPageToken fields on the request are ignored.
+func (c *NexusIncomingServiceClient) listAndFilterByName(
+	ctx context.Context,
+	request *operatorservice.ListNexusIncomingServicesRequest,
+) (*operatorservice.ListNexusIncomingServicesResponse, error) {
+	nameIssues := c.getServiceNameIssues(request.Name)
+	if err := nameIssues.GetError(); err != nil {
+		return nil, err
+	}
+
+	result := &operatorservice.ListNexusIncomingServicesResponse{Services: []*nexus.IncomingService{}}
+	pageSize := c.config.listDefaultPageSize()
+	var currentPageToken []byte
+
+	for ctx.Err() == nil {
+		resp, err := c.persistence.ListNexusIncomingServices(ctx, &p.ListNexusIncomingServicesRequest{
+			LastKnownTableVersion: 0,
```

Comment:
Yes, I confirmed.

Reply:
I made an assumption in the cassandra store implementation that `LastKnownTableVersion=0` always meant get first page. I've updated the logic so that if table version is 0 and next page token is not nil, it will get the indicated page and not do the table version validation.

Reply:
Yeah, we were being too strict before with regards to pagination. I'm glad we relaxed the requirements, it'll make for better UX.